linked_list/linked_list.py

```diff
@@ -1,102 +1,183 @@
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def search(self, value):
-        pass
+        current = self.head
+
+        while current != None:
+            if current.value == value:
+                return True
+            current = current.next
+
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def length(self):
-        pass
+        counter = 0
+        current  = self.head
+        
+        while current != None:
+            current = current.next
+            counter += 1
+        
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_at_index(self, index):
-        pass
+        if self.length() < index:
+            return None
+
+        current = self.head
+        current_index = 0
+        while current != None:
+            if current_index == index:
+                return current.value
+            current = current.next
+            current_index += 1
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        while current.next != None:
+            current = current.next
+
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head == None:
+            self.head = new_node
+            return 
+
+        current = self.head
+        while current.next != None:
+            current = current.next
+        
+        current.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        max_value = self.head.value
+        current = self.head.next
+
+        while current != None:
+            if max_value < current.value:
+                max_value = current.value
+            current = current.next
+
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def delete(self, value):

```

Comment:
👍 But since you have to find the node to delete it, the time complexity of this method is O(n)